README
# termcast

This module is a client for the L<http://termcast.org/> service, which allows
broadcasting of a terminal session for remote viewing. It can be installed
either via `pip install termcast_client` or by downloading the standalone
binary
[here](https://raw.githubusercontent.com/doy/python-termcast-client/master/termcast).

To run it, just run `termcast`. A new shell will open, and it will start
streaming data to the termcast server. You can view the contents by following
the instructions at [termcast.org](http://termcast.org/).

## Detailed Usage

    termcast [--host <host>] [--port <port>] [--username <username>]
             [--password <password>] [command...]

* `--host` defaults to `noway.ratry.ru`
* `--port` defaults to `31337`
* `--username` defaults to the contents of `$USER`
* `--password` defaults to asdf
* the command defaults to the contents of `$SHELL`, or `/bin/sh` if not set


Changes
Revision history for termcast_client

0.1.3  2014-10-22
    - Add a socket timeout (defaulting to 5 seconds)

0.1.2  2014-10-17
    - Reconnect to the server when the connection drops

0.1.1  2014-10-16
    - Python 2 support (Tom Ballinger)
    - Preliminary SSL support (for servers that support it). This is not yet
      documented and may change at any point as I work out the details.


0.1.0  2014-09-23
    - Initial revision


termcast_client.py
import argparse
import json
import os
import pity
import shutil
import signal
import socket
import sys

class Client(object):
    def __init__(self, host, port, username, password):
        self.host = host
        self.port = port
        self.username = username
        self.password = password

    def run(self, argv):
        sock = socket.socket()
        sock.connect((self.host, self.port))
        sock.send(self._build_connection_string())
        self.winch_set = False
        pity.spawn(
            argv,
            lambda fd: self._master_read(fd, sock),
            handle_window_size=True
        )

    def _master_read(self, fd, sock):
        if not self.winch_set:
            prev_handler = signal.getsignal(signal.SIGWINCH)
            signal.signal(
                signal.SIGWINCH,
                lambda signum, frame: self._winch(
                    sock,
                    lambda: prev_handler(signum, frame)
                )
            )
            self.winch_set = True

        data = os.read(fd, 1024)
        sock.send(data)
        return data

    def _winch(self, sock, prev_handler):
        prev_handler()
        # XXX a bit racy - should try to avoid splitting existing escape
        # sequences
        sock.send(self._build_winsize_metadata_string())

    def _build_connection_string(self):
        auth = (
            b'hello ' +
            self.username.encode('utf-8') +
            b' ' +
            self.password.encode('utf-8') +
            b'\n'
        )
        metadata = self._build_connection_metadata_string()
        return auth + metadata

    def _build_connection_metadata_string(self):
        # for now
        return self._build_winsize_metadata_string()

    def _build_winsize_metadata_string(self):
        size = shutil.get_terminal_size()
        return self._build_metadata_string({
            "geometry": [ size.columns, size.lines ],
        })

    def _build_metadata_string(self, data):
        return b'\033]499;' + json.dumps(data).encode('utf-8') + b'\007'

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--host', default="noway.ratry.ru")
    parser.add_argument('--port', type=int, default=31337)
    parser.add_argument('--username', default=os.getenv("USER"))
    parser.add_argument('--password', default="asdf")
    parser.add_argument(
        'command',
        nargs=argparse.REMAINDER,
    )

    args = parser.parse_args(sys.argv[1:])

    client = Client(
        host=args.host,
        port=args.port,
        username=args.username,
        password=args.password,
    )

    command = args.command
    if len(args.command) < 1:
        command = os.getenv("SHELL", default="/bin/sh")
    client.run(command)


termcast_client/pity.py
import errno
import fcntl
import os
import pty
import signal
import termios
import tty

from . import py2compat

CHILD = pty.CHILD
STDIN_FILENO = pty.STDIN_FILENO
STDOUT_FILENO = pty.STDOUT_FILENO
STDERR_FILENO = pty.STDERR_FILENO

def fork(handle_window_size=False):
    # copied from pty.py, with modifications
    master_fd, slave_fd = openpty()
    slave_name = os.ttyname(slave_fd)
    pid = os.fork()
    if pid == CHILD:
        # Establish a new session.
        os.setsid()
        os.close(master_fd)

        if handle_window_size:
            clone_window_size_from(slave_name, STDIN_FILENO)

        # Slave becomes stdin/stdout/stderr of child.
        os.dup2(slave_fd, STDIN_FILENO)
        os.dup2(slave_fd, STDOUT_FILENO)
        os.dup2(slave_fd, STDERR_FILENO)
        if (slave_fd > STDERR_FILENO):
            os.close (slave_fd)

        # Explicitly open the tty to make it become a controlling tty.
        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
        os.close(tmp_fd)
    else:
        os.close(slave_fd)

    # Parent and child process.
    return pid, master_fd, slave_name


def openpty():
    return pty.openpty()

def spawn(argv, master_read=pty._read, stdin_read=pty._read, handle_window_size=False):
    # copied from pty.py, with modifications
    # note that it references a few private functions - would be nice to not
    # do that, but you know
    if type(argv) == type(''):
        argv = (argv,)
    pid, master_fd, slave_name = fork(handle_window_size)
    if pid == CHILD:
        os.execlp(argv[0], *argv)
    try:
        mode = tty.tcgetattr(STDIN_FILENO)
        tty.setraw(STDIN_FILENO)
        restore = 1
    except tty.error:    # This is the same as termios.error
        restore = 0

    if handle_window_size:
        signal.signal(
            signal.SIGWINCH,
            lambda signum, frame: _winch(slave_name, pid)
        )

    while True:
        try:
            pty._copy(master_fd, master_read, stdin_read)
        except OSError as e:
            if e.errno == errno.EINTR:
                continue
            if restore:
                tty.tcsetattr(STDIN_FILENO, tty.TCSAFLUSH, mode)
        break

    os.close(master_fd)
    return os.waitpid(pid, 0)[1]

def clone_window_size_from(slave_name, from_fd):
    slave_fd = os.open(slave_name, os.O_RDWR)
    try:
        fcntl.ioctl(
            slave_fd,
            termios.TIOCSWINSZ,
            fcntl.ioctl(from_fd, termios.TIOCGWINSZ, " " * 1024)
        )
    finally:
        os.close(slave_fd)

def _winch(slave_name, child_pid):
    clone_window_size_from(slave_name, STDIN_FILENO)
    os.kill(child_pid, signal.SIGWINCH)


termcast_client/py2compat.py
import fcntl
import os
import struct
import sys
import termios

from collections import namedtuple

terminal_size = namedtuple('terminal_size', ['columns', 'lines'])

py2 = sys.version_info[0] == 2

def get_terminal_size(fallback=(80, 24)):
    """Based on Python 3 os.get_terminal_size()"""
    # columns, lines are the working values
    try:
        columns = int(os.environ['COLUMNS'])
    except (KeyError, ValueError):
        columns = 0

    try:
        lines = int(os.environ['LINES'])
    except (KeyError, ValueError):
        lines = 0

    # only query if necessary
    if columns <= 0 or lines <= 0:
        try:
            winsize = fcntl.ioctl(sys.__stdout__.fileno(),
                                  termios.TIOCGWINSZ, '\000' * 8)
            size = terminal_size(*struct.unpack('hhhh', winsize)[1::-1])

        except (NameError, OSError):
            size = terminal_size(*fallback)
        if columns <= 0:
            columns = size.columns
        if lines <= 0:
            lines = size.lines

    return terminal_size(*size)


termcast_client/__init__.py
import argparse
import hashlib
import json
import os
import shutil
import signal
import socket
import ssl
import sys
import time

from . import pity
from . import py2compat

class Client(object):
    def __init__(self, host, port, username, password, timeout, tls, fingerprint):
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.timeout = timeout
        self.tls = tls
        self.fingerprint = fingerprint

    def run(self, argv):
        self._new_socket()
        self.winch_set = False
        pity.spawn(
            argv,
            self._master_read,
            handle_window_size=True
        )

    def _new_socket(self):
        self.sock = socket.socket()
        self.sock.settimeout(self.timeout)
        self.sock.connect((self.host, self.port))
        if self.tls:
            self._starttls()
        self.sock.send(self._build_connection_string())

    def _renew_socket(self):
        self.sock.close()

        while True:
            print("Disconnected from server, reconnecting...")
            time.sleep(5)
            try:
                self._new_socket()
                return
            except:
                pass

    def _starttls(self):
        self.sock.send(b'starttls\n')
        context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
        if self.fingerprint is not None:
            context.check_hostname = False
            context.verify_mode = ssl.CERT_NONE
        self.sock = context.wrap_socket(self.sock, server_hostname=self.host)
        if self.fingerprint is not None:
            remote = hashlib.sha1(self.sock.getpeercert(True)).hexdigest()
            if remote != self.fingerprint:
                raise Exception("Invalid fingerprint received: %s" % remote)

    def _master_read(self, fd):
        if not self.winch_set:
            self.prev_handler = signal.getsignal(signal.SIGWINCH)
            signal.signal(signal.SIGWINCH, self._winch)
            self.winch_set = True

        data = os.read(fd, 1024)
        to_send = data
        datalen = len(to_send)
        while datalen > 0:
            try:
                sent = self.sock.send(to_send)
                if sent == 0:
                    raise "disconnect"
                to_send = to_send[sent:]
                datalen = len(to_send)
            except:
                self._renew_socket()

        return data

    def _winch(self, signum, frame):
        self.prev_handler(signum, frame)
        # XXX a bit racy - should try to avoid splitting existing escape
        # sequences
        self.sock.send(self._build_winsize_metadata_string())

    def _build_connection_string(self):
        auth = (
            b'hello ' +
            self.username.encode('utf-8') +
            b' ' +
            self.password.encode('utf-8') +
            b'\n'
        )
        metadata = self._build_connection_metadata_string()
        return auth + metadata

    def _build_connection_metadata_string(self):
        # for now
        return self._build_winsize_metadata_string()

    def _build_winsize_metadata_string(self):
        if py2compat.py2:
            size = py2compat.get_terminal_size()
        else:
            size = shutil.get_terminal_size()
        return self._build_metadata_string({
            "geometry": [ size.columns, size.lines ],
        })

    def _build_metadata_string(self, data):
        return b'\033]499;' + json.dumps(data).encode('utf-8') + b'\007'

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--host', default="noway.ratry.ru")
    parser.add_argument('--port', type=int, default=31337)
    parser.add_argument('--username', default=os.getenv("USER"))
    parser.add_argument('--password', default="asdf")
    parser.add_argument('--timeout', default=5)
    parser.add_argument('--tls', action='store_true')
    parser.add_argument('--fingerprint')
    parser.add_argument(
        'command',
        nargs=argparse.REMAINDER,
    )

    args = parser.parse_args(sys.argv[1:])

    client = Client(
        host=args.host,
        port=args.port,
        username=args.username,
        password=args.password,
        timeout=args.timeout,
        tls=args.tls,
        fingerprint=args.fingerprint,
    )

    command = args.command
    if len(args.command) < 1:
        command = os.getenv("SHELL", default="/bin/sh")
    client.run(command)
